Design/3093-graphql-defer.md

```diff
@@ -0,0 +1,356 @@
+
+The data being retained and combined should not require another pass through the GraphQL executor though.
+
+### Completion handler
+
+`GraphQLResult` should be modified to provide query completion blocks with a high-level abstraction of whether the request has been fulfilled or is still in progress. This prevents clients from having to dig into the deferred fragments to identify the state of the overall request.
+
+Potential solutions:
+1. Introduce a new property on the `GraphQLResult` type that can be used to express the state of the request.
+
+```swift
+// New Response type and property
+public struct GraphQLResult<Data: RootSelectionSet> {
+  // other properties and types not shown
+
+  public enum Response {
+    case partial
+    case complete
+  }
+
+  public let response: Response
+}
+
+// Sample usage in an app completion block
+client.fetch(query: ExampleQuery()) { result in
+  switch (result) {
+  case let .success(data):
+    switch (data.response) {
+    case .complete:
+    case .partial:
+    }
+  case let .failure(error):
+  }
+}
+```
+
+2. Another way which may be a bit more intuitive is to make the `server` case on `Source` have an associated value since `cache` sources will always be complete. The cache could return partial responses for deferred operations but for the initial implementation we will probably only write the cache record once all deferred fragments have been received. This solution becomes invalid though once the cache can return partial responses, with that in mind maybe option 1 is better.
+
+```swift
+// Updated server case on Source with associated value of Response type
+public struct GraphQLResult<Data: RootSelectionSet> {
+  // other properties and types not shown
+
+  public enum Response {
+    case partial
+    case complete
+  }
+
+  public enum Source: Hashable {
+    case cache
+    case server(_ response: Response)
+  }
+}
+
+// Sample usage in an app
+client.fetch(query: ExampleQuery()) { result in
+  switch (result) {
+  case let .success(data):
+    switch (data.source) {
+    case .server(.complete):
+    case .server(.partial):
+    case .cache:
+    }
+  case let .failure(error):
+  }
+}
+```
+
+## GraphQL execution
+
+The executor currently executes on an entire operation selection set. It will need to be adapted to be able to execute on a partial response when deferred fragments have not been received and on an isolated fragment selection set so that incremental responses can be parsed alone instead of needing to execute on the whole operation’s selection set.
+
+An alternative to parsing isolated fragment selection sets would be to execute on all the data currently received. The inefficiency with this is executing on data that would have already been executed on during prior responses.
```

Comment:
Yeah, I'm skeptical that we're going to be able to get partial execution to work without a whole lot of new bugs in a reasonable amount of time. I'm sure it's possible, but I'm not sure if it's worth attempting for this 1st iteration on defer.